enhancements/20240604-lvm-csi-driver.md

```diff
@@ -0,0 +1,77 @@
+
+#### The user creates the PCI pass-through VM and wants a higher performance.
+When users use the features with PCI pass-through, it means they will redirect the VM to a specific host. Also, the user wants a higher-performance disk for the VM. The LVM CSI driver could help users create the PVC with an LVM volume on a local disk. That could help users have better performance and latency.
+
+### User Experience In Detail
+
+The user creates a VM and starts with the feature with PCI pass-through. Then, the user creates the PVC with the LVM volume. The user attaches the PVC to the VM through hotplug and uses the LVM volume in the VM.
+
+### API changes
+TBD
+
+## Design
+
+### Implementation Overview
+
+A overview of the design could be checked below:
+![lvm-csi-driver-arch](20240604-lvm-csi-driver/lvm-csi-driver-arch.png)
+
+### LVM CSI Driver
+At first, we consider two existing solutions: [metal-stack/lvm-csi-driver](https://github.com/metal-stack/csi-driver-lvm) and [topolvm/topolvm](https://github.com/topolvm/topolvm). Both solutions support the CSI topology feature, and other features like volume capability and snapshot are only supported in the topolvm. However, the topolvm is more complex than the lvm-csi-driver. We could try to add the capability and snapshot feature to the lvm-csi-driver and keep it as simple as we want.
+
+So, we chose the [metal-stack/lvm-csi-driver](https://github.com/metal-stack/csi-driver-lvm) as the base LVM CSI driver. This project is lightweight and already has the essential capability for our requirements. We want to introduce it in the first version. Then, we can add more features based on the requirements.
+
+The LVM CSI driver includes the CSI plugin and the LVM provisioner. The CSI plugin handles all CSI requests, while the LVM provisioner manages LVM operations such as creation, deletion, expansion, and snapshotting. This division of responsibilities ensures efficient and effective management of our storage infrastructure.
+
+In the initial version, we propose a significant change that will enhance user flexibility. We plan to allow users to predefine the volume group. The provisioner will then use this predefined volume group and logical volume type (e.g., mirror, stripe) to create the volume. This change will enable users to group similar types of disks together, thereby ensuring optimal performance and capability.
+
+With this change, the StorageClass will now contain two additional parameters: `vgName` and `lvType`. The `vgName` refers to the volume group name, and the `lvType` specifies the logical volume type. These parameters are essential for the LVM provisioner to create the volume according to the user's specifications.
+
+### Node Disk Manager
+Although we can create a volume group manually, the node disk manager can help us do so. The node disk manager will create the volume group based on the disks the user chooses. Then, the user can use the volume group name to create StorageClass and PVC.
```

Comment:
How does the user tell NDM which disks to use to create the VG?

Reply:
I gave a link to the NDM refactor, which contains more details.